refactored

jonschlinkert opened the pull request:
Not ready to merge, just wanted to get this pushed up so we can discuss. we can also take another look at https://github.com/doowb/composer/issues/30 before merging.

jonschlinkert committed:
simplify matcher, since only basic globs are allowed

jonschlinkert committed:
paralle/series

jonschlinkert committed:
clean up

jonschlinkert committed:
simplify Task getters

jonschlinkert committed:
clean up

doowb commented on lib/task.js:
```diff
@@ -1,155 +1,121 @@
+      case 'undefined':
+        return false;
+      case 'boolean':
+        return opts.skip === false;
```
I think this should be `return opts.skip === true`

doowb commented:
This is awesome! I think having the stack of running tasks will help with #30

jonschlinkert committed:
adds examples

jonschlinkert committed:
add build timers

jonschlinkert committed:
use Emitter

jonschlinkert committed:
use `Map` for payload and `tasks`

jonschlinkert committed:
use utils, custom emitter

Since `.create` allows the user to pass a custom emitter, we're using a simple built-in emitter to reduce deps and provide a good baseline experience. If you need more, use a custom emitter.

jonschlinkert committed:
run verb to generate readme

doowb commented on lib/task.js:
```diff
@@ -1,155 +1,123 @@
-const define = require('define-property');
-const iterator = require('./iterator');
+
+const Emitter = require('events');
```
Should this be `const Emitter = require('./events');`?

doowb commented on lib/task.js:
```diff
@@ -1,155 +1,123 @@
-const Emitter = require('@sellside/emitter');
-const define = require('define-property');
-const iterator = require('./iterator');
+
```
Extra line!!!

doowb commented on lib/task.js:
```diff
@@ -1,155 +1,123 @@
+
+          if (this.callback.length === 0) {
+            if (res.then) {
+              res.then(() => finish());
```
This line could be `res.then(val => finish(null, val));` so if any values are returned from the promise they'll be passed through like with callbacks.

doowb commented on lib/tasks.js:
```diff
@@ -0,0 +1,397 @@
+ * @api public
+ */
+
+module.exports = Emitter => {
```
Should this be validated or at least default to the internal emitter: `module.exports = (Emitter = require('./events')) => {`

jonschlinkert replied:
No, `task` should not have a customizable emitter. We need to make sure that task doesn't get, say, `Templates` or `Assemble` as the Emitter, which would cause major issues.

doowb commented on lib/tasks.js:
```diff
@@ -0,0 +1,397 @@
+      super(options);
+      this.options = options;
+      this.taskStack = new Set();
+      this.taskState = new Map();
```
I don't see this used anyplace other than in `resolveBuild`. Is this so tasks have something to add state to?

jonschlinkert replied:
> Is this so tasks have something to add state to?

yes, but I think we should remove it if you haven't already

doowb commented on lib/tasks.js:
```diff
@@ -0,0 +1,397 @@
+            continue;
+          }
+
+          const run = task.run(options);
```
I think the event listeners on the next 2 lines should come before this line. Since `task.run` emits the `starting` event, that won't get caught by this listener.

doowb commented on lib/tasks.js:
```diff
@@ -0,0 +1,397 @@
+
+          const run = task.run(options);
+          task.on('finished', () => stack.delete(task));
+          task.on('starting', () => stack.add(task));
```
Should these be `task.once`? Otherwise, if a task is executed multiple times (like when using a watch task) I think it will cause a memory leak since `stack` will never be released from memory.

doowb commented on lib/tasks.js:
```diff
@@ -0,0 +1,397 @@
+
+          const run = task.run(options);
+          task.on('finished', () => stack.delete(task));
+          task.on('starting', () => stack.add(task));
```
Same event comments as in `series`.

doowb commented on package.json:
```diff
@@ -19,49 +19,30 @@
   "main": "index.js",
   "engines": {
-    "node": ">=8"
+    "node": ">=4"
```
Need to bump this since `async/await` isn't available until node 8.

doowb commented:
@jonschlinkert I added review comments to the code inline. Looking good!

doowb committed:
clean up after code review

doowb commented:
@jonschlinkert I pushed up the changes based on my review. I think this looks good if you want to merge and publish (unless there's more that you want to add).

jonschlinkert commented on lib/tasks.js:
```diff
@@ -20,7 +20,7 @@ const noop = cb => cb();
  */
 
-module.exports = Emitter => {
+module.exports = (Emitter = require('./events')) => {
```
this should be moved up to the top

jonschlinkert commented on lib/tasks.js:
```diff
@@ -368,10 +367,10 @@ module.exports = Emitter => {
+    promise.then(val => cb(null, val)).catch(cb);
     return;
   }
-  return promise.then(() => app.taskState);
```
if no callback is passed, and the user does not return a value in the promise, will this still ensure that the state is returned? I didn't look at changes in the whole file yet.

jonschlinkert committed:
add generator-plugin support

doowb committed:
move default event emitter require to top of files

jonschlinkert committed:
remove comment

update `.task` to use `.createOptions`

jonschlinkert committed:
remove kind-of

Also use `Object.create(null)`. This probably won't make much difference, since `Object.assign()` only assigns own properties, but in principle there could be performance improvements.

jonschlinkert committed:
remove unused case, per @doowb
